Add stream converters for responses

NathanReb opened the pull request:
Fixes #21.

This PR adds `to_stream` functions to convert `('a, _) Response.t` to `'a stream`.
The current implementation uses the following stream types:
- Blocking drivers use `'a Seq.t`, this adds a dependency over the compatibility package `seq` to allow `ocaml < 4.07`
- Lwt drivers use `'a Lwt_stream.t`
- Async drivers use `'a Async.Pipe.Reader.t`

The converter returns an `('a, retrieval_error) result` stream. If we get an error trying to retrieve a row, the error is added to the stream as its last element. That allows the consumer of the stream to properly deal with the error.

NathanReb commented:
This is still a first draft, but I'd love to hear your feedback!

I'd like to add tests but I'm unsure what the best place would be. My first guess was `test_sql.ml` but that actually seems a bit too high level for that kind of tests, any pointers or suggestions welcome!

I had to make a choice regarding error handling. While it seems like a reasonable behaviour I'd gladly change it if you have something better in mind.

paurkedal commented on lib-blocking/caqti_blocking.ml:
```diff
@@ -42,6 +42,21 @@ module System = struct
+          state := new_state;
+          Seq.Cons (a, wrapped_f)
+      in
+      wrapped_f
```
This can be simplified to
```ocaml
    let rec from_fun f state () =
      (match f state with
       | None -> Seq.Nil
       | Some (a, state') -> Seq.Cons (a, from_fun f state'))
```

paurkedal commented on lib-driver/caqti_driver_mariadb.ml:
```diff
@@ -374,6 +375,16 @@ module Connect_functor (System : Caqti_driver_sig.System_unix) = struct
+      let to_stream {query; res; row_type} =
+        let f got_error =
+          if got_error then return None else
+          decode_next_row ~query res row_type >>= 
```
Please remove whitespace at end of line.

paurkedal commented on lib/caqti_response_sig.mli:
```diff
@@ -79,4 +81,9 @@ module type S = sig
+  val to_stream : ('b, 'm) t -> ('b, [> Caqti_error.retrieve] as 'e) result stream
+  (** [to_stream resp] returns a stream whose elements are the decoded rows returned
+      by [resp].
+      If an error is encountered it will be the last element of the stream. *)
```
Please break lines before column 81, for the signature I prefer right after the colon in this case.

paurkedal commented:
> I had to make a choice regarding error handling. While it seems like a reasonable behaviour I'd gladly change it if you have something better in mind.

This was the first think I noticed, and it would have been good to have non-exception error handling built into the streams here, but I think it is a good choice and I think the best we can do given the choice of Caqti not to use exceptions for runtime errors.

I don't know async much, but `'a Async.Pipe.Reader.t` is nice and I think i good choice here.  As you can see from my review, `Seq` in fact also uses the precise same `state -> value * state` pattern.

As for testing, I have put most in `test_sql.ml`, but that file is unified between the different concurrency models.  In this case it is easier to write different code for each variant?

paurkedal commented:
I am considering whether it would be worth our own error-handling stream
```ocaml
module Stream = struct
  type ('a, 'e) t = unit -> ('a, 'e) node future

  and (+'a, +'e) node =
    | Nil
    | Cons of 'a * ('a, 'e) t
    | Error of 'e
end
```
Maybe provided though `Caqti_connect_sig.S` similar to `Pool`.  This would avoid wrapping elements in the `result` type, and avoid extending `System_common`.  I would consider reimplementing other iterators in terms of this to simplify driver code.  This doesn't directly provide the standard iterators from the other libraries, but given that they don't fit perfectly for our purpose, isn't it tidier to use this a base?

NathanReb commented:
I'm a bit torn about this. While I definitely see why a custom stream could improve the error handling here and spare us the need to deal with `result` wrapped values, I'm not sure it's a good idea. The above mentioned libraries have their own set of functions that people are used to work with and re-implementing all of them probably isn't `Caqti`'s responsibility. I think users of this new API (myself included) would like to be able to compose those streams and it obviously won't be as convenient if `Caqti` comes with its own.

IMHO, there's two way we could implement streams within `caqti`:
- We define a really safe but limited API to keep the error handling clean. That means we lose the ability to compose them with regular streams, modules that shouldn't will now have to depend on `Caqti` and we have to maintain a stream library.
- We define a looser API to allow people to eventually convert those streams back to the ones mentioned above, which they will do most of the time, losing the cleanliness of it's error handling on the way. That's pretty much equivalent to what we have now.

I might be wrong and miss something here of course but it does feel a little odd to redefine a stream module here.

Let me know what you think!

NathanReb commented:
I just added tests, turned out to be fairly easy to add them to `test_sql.ml` so I went for it, let me know if that suits you!

paurkedal commented:
I do understand your concern.  I would not suggest defining the stream type here if there was a good option available.  The two missing pieces are integration of error handing and a cross-concurrency alternative which would allow also using the stream internally in Caqti to keep the driver code as simple as possible.  Strictly speaking drivers would only need to define a `fetch_row` function, and `to_stream` is just a fancy variant of that.

An relevant consideration is how these streams will be used.  I think the main use case will be to write custom code inside the call-back which would pick off and process elements one by one in a nontrival way.  If on the other hand one wants to pass streams to existing libraries, then matching our choice with corresponding streams may seem preferred, but would one not already need to transform the stream to get rid of the `result`, anyway?  That is, I suspect the wrapping already makes the stream effectively incompatible with existing libraries, which may expect to receive errors as excpections if supported.

paurkedal commented:
Thanks for the test, I'll take a look at it.  And sorry for raising the issue of the stream format only after you sent the PR, but I think it is an important decision.

paurkedal commented:
Just one more though on the choice of stream.  Independent of what mechanism is provided in `Caqti_response_sig.S` by the driver, it should be possible to add different convenience functions in `Caqti_async` and `Caqti_lwt`.  Those modules don't need to restrict themselves to `Caqti_connect_sig.S`, though if we go that route, we'd need to include the whole signature in order to change the `CONNECTION` signature.

paurkedal commented:
The test looks good, that should suffice.

NathanReb commented:
>  And sorry for raising the issue of the stream format only after you sent the PR, but I think it is an important decision.

No problem! I agree this is indeed an important decision and shouldn't be rushed.

Being able to use those streams internally is indeed a good argument in favor of such custom streams.

You also have a very good point about the need for conversion anyway. I think as long as there's a way to convert those streams to `Lwt`, `Async` or `Seq` ones, a custom type is fine because that would both be convenient for users and spare us the need to expose the whole range of helper functions on `caqti`'s stream.

paurkedal commented:
Good, I think we agree lwt and async specific streams are best implemented separately, to allow accounting for differences.  I'll have a look at the needed refactoring in the weekend if not before.

This allows us to rely on lwt-exceptions for the Lwt part.  Likewise, the blocking implementation can use regular exceptions.  But, as far as I read the Async pipes, there seems to be no way for the writer to signal an error to the reader without wrapping the error into the pipe elements.

NathanReb commented:
Just so you know, I started working on a `Caqti_stream` functor that would implement `fold`, `fold_s` and `iter_s` so that the driver code could be simplified. The stream type would be exposed so that we can easily convert them to the `Lwt`, `Async` and `Seq` variants.

Is that aligned with what you were suggesting?

NathanReb commented:
I pushed two new commits with what I currently have.

It adds a `Caqti_stream` functor which is used to build `Response.Stream` and that's what is used in `Response.to_stream`.
The next steps would be to add converters in `Caqti_blocking`, `Caqti_lwt` and `Caqti_async` and eventually to express `fold`, `fold_s`, `iter_s`, `collect_rev_list` and `collect_list` using `Caqti_stream`.

I'd love to hear your feedback before going further on this. Especially, I'm not sure `Response` is the best place for the `Stream` module to sit in. It should probably be built higher up and passed to the right functors so that `Response` and `Connect` can use it.

Let me know what you think!

paurkedal commented:
Nice!  This is very close to what I had in mind.

Yes, I think we should consider how to expose a connection-independent `Stream`.  If we made the instances path-equal to `Caqti_stream.Make (Lwt)` and `Caqti_stream.Make (Async.Deferred)`, then it would be less important where it was defined.  Since there is no common form for the map operation, this would incur an inefficiency for `to_list` and maybe later additions, so maybe it's better to define them in lib-lwt and lib-async and pass the modules, as you suggest.   I haven't considered exactly how to do it.

I was originally undecided about labelling the `~f` argument in iterators, but after sketching a bit I convinced myself that this is a sufficiently clear improvement that it's worth deviating from the common signatures.  (And thanks for not labelling the `'state` argument.)

paurkedal commented:
I think it is better not to expose the more precise `Caqti_stream.Make (...)` paths for now.  Having the stream implementations at the `Caqti_lwt` and `Caqti_async` level would be nice, but maybe not strictly required.  My reasoning is that we are not providing a stream implementation as much as a canonical way to expose the rows from the response, and from a backwards compatibility point of view it's easier to add module equalities than to remove them.

You don't need to rewrite `fold_s`, `iter_s` and `fold` for this PR unless you wish, there mere opportunity of simplifying the drivers later is already appreciated.

I was wondering, did you choose the `Err` constructor name to avoid clash with `Error`?  I think we could use `Error` since I expect it will always be qualified in client code.  (I am thinking a bit about what the most likely constructor name would be if this was added as a separate library or proposed for stdlib, though the hope of a type equality for our stream later may be a bit far fetched.)

paurkedal commented:
I pushed a commit which moves the stream out of `Response` to your branch.  Feel free to reset if you have a better idea.

NathanReb commented:
I did not get a chance to get back at that yet but that's pretty much what I had in mind from your last comments, I'll try to get some time to hopefully wrap this up this afternoon!

paurkedal commented:
Before I merge anything, it's best if you add a copyright header to the new files.  Please use the LGPL-with-linking-exception text from another file, and feel free to put yourself as the copyright holder.  With your permission I'd also add you as an author to the opam files.

paurkedal commented:
Do you have any thoughts on how to deal with errors if we provide lwt,async-native streams?

NathanReb commented:
No unfortunately haven't had time to look into it today but that's on my TO-DO list!

paurkedal commented:
Sure, there is no hurry.

NathanReb commented:
I just spent some time trying to move the Stream module definition to `Caqti_blocking`, `Caqti_lwt` and `Caqti_async` so that I could define functions relying on the corresponding libraries in the Stream module itself as I believed it would make for a better API and then pass it as a functor argument.

After struggling with all the functors and module types for far too long I decided to let go of it for now so I added conversion function in the form `stream_to_<regular_stream_type>` in `Caqti_blocking` and `Caqti_lwt`. I didn't look into the Async version but it is not strictly necessary and can be added later.
Let me know what you think about it.

I added the copyright headers (in your name, doesn't need to be mine there).

I also renamed the `Err` constructor to `Error` in `Caqti_stream`, the compilers requires a few hints there to distinguish the `result` constructor from the `node` one. That was the reason why I used a different name in the first place but we can work around it. I used type constraints but if you have any better idea I'll gladly change it.

NathanReb commented:
It might be worth considering adding `catch` like functions to wrap the use of streams and get an `('a, Caqti_error.t) result` back as with those new streams users might end up re-implementing such functions on their own.

What I suggest is for instance to add the following to `Caqti_lwt`:
```ocaml
let catch f =
  Lwt.catch
    (fun () -> f () >>= Lwt.return_ok)
    (function
      | Caqti_error.Exn err -> Lwt.return_error err
      | exn -> Lwt.fail exn)
```
That's not strictly required/can be done later but I just wanted to get your opinion on the matter.

paurkedal commented:
Thanks for the adjustments.  It looks good now except for the stream conversion.

Adding some minimal exception handling for callbacks would make the library more resilient against bugs, but I'm much less found of introducing exceptions as a normal mode of error handling within `call`.  My original though was to amend dedicated helper in `CONNECTION` for lwt and blocking for the purpose, though as you have noticed, the implementation is not straight forward: we would have to write custom `connect` and `connect_pool` functions either duplicating code from `Caqti_connect` or exposing some of its internals, so it will have some impact on maintainability.

paurkedal commented:
Would you mind if we merge this without stream conversions for now?

NathanReb commented:
Nope not at all! I was actually about to suggest that. I'll remove them!

NathanReb commented:
There you go!

paurkedal commented:
Thanks!  I quick fix from me, then ready to merge if the just enabled CI says so.

paurkedal commented:
Close enough. I broke compatibility with 4.05, will fix that afterwards.

paurkedal commented:
@NathanReb I will make a release very soon including the stream API.  Sorry for the long wait. @leamingrad and I made some changes to it, so let me know if you have any final comments before fixing the API for the release.
